File: gems/server/ext/itsi_server/src/server/middleware_stack/middlewares/log_requests.rs
use async_trait::async_trait;
use either::Either;
use itsi_tracing::*;
use magnus::error::Result;
use serde::Deserialize;

use crate::server::itsi_service::RequestContext;
use crate::server::types::{HttpRequest, HttpResponse};

use super::string_rewrite::StringRewrite;
use super::{FromValue, MiddlewareLayer};

/// Logging middleware for HTTP requests and responses
///
/// Supports customizable log formats with placeholders
#[derive(Debug, Clone, Deserialize)]
pub struct LogRequests {
    pub before: Option<LogConfig>,
    pub after: Option<LogConfig>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct LogConfig {
    level: LogMiddlewareLevel,
    format: StringRewrite,
}

#[derive(Debug, Clone, Deserialize)]
pub enum LogMiddlewareLevel {
    #[serde(rename(deserialize = "INFO"))]
    Info,
    #[serde(rename(deserialize = "TRACE"))]
    Trace,
    #[serde(rename(deserialize = "DEBUG"))]
    Debug,
    #[serde(rename(deserialize = "WARN"))]
    Warn,
    #[serde(rename(deserialize = "ERROR"))]
    Error,
}

impl LogMiddlewareLevel {
    pub fn log(&self, message: String) {
        match self {
            LogMiddlewareLevel::Trace => trace!(message),
            LogMiddlewareLevel::Debug => debug!(message),
            LogMiddlewareLevel::Info => info!(message),
            LogMiddlewareLevel::Warn => warn!(message),
            LogMiddlewareLevel::Error => error!(message),
        }
    }
}

#[async_trait]
impl MiddlewareLayer for LogRequests {
    async fn initialize(&self) -> Result<()> {
        Ok(())
    }

    async fn before(
        &self,
        req: HttpRequest,
        context: &mut RequestContext,
    ) -> Result<Either<HttpRequest, HttpResponse>> {
        context.track_start_time();
        if let Some(LogConfig { level, format }) = self.before.as_ref() {
            level.log(format.rewrite_request(&req, context));
        }

        Ok(Either::Left(req))
    }

    async fn after(&self, resp: HttpResponse, context: &mut RequestContext) -> HttpResponse {
        if let Some(LogConfig { level, format }) = self.after.as_ref() {
            level.log(format.rewrite_response(&resp, context));
        }

        resp
    }
}

impl FromValue for LogRequests {}

File: gems/server/lib/itsi/server/config/middleware/log_requests.rb
module Itsi
  class Server
    module Config
      class LogRequests < Middleware

        insert_text <<~SNIPPET
        log_requests \\
          before: { level: ${1|"INFO","WARN","ERROR","DEBUG"|} format: ${2|"[{request_id}] {method} {path_and_query} - {addr} "|} },
          after: { level: ${3|"INFO","WARN","ERROR","DEBUG"|}, format: ${4|"[{request_id}] └─ {status} in {response_time}"|} }
        SNIPPET

        detail "Enable logging before or after requests"

        LogRequestConfig = TypedStruct.new do
          {
            level: (Required() & Enum(%w[INFO WARN ERROR DEBUG])).default("INFO"),
            format: Type(String)
          }
        end

        schema do
          {
            before: Type(LogRequestConfig),
            after: Type(LogRequestConfig)
          }
        end

      end
    end
  end
end
